modules/javafx.base/src/main/java/com/sun/javafx/collections/ObservableListWrapper.java

```diff
@@ -169,40 +169,36 @@ public void remove(int fromIndex, int toIndex) {
 
     @Override
     public boolean removeAll(Collection<?> c) {
+        if (this.isEmpty() || c.isEmpty()) {
+            return false;
+        }
         beginChange();
-        BitSet bs = new BitSet(c.size());
-        for (int i = 0; i < size(); ++i) {
+        boolean removed = false;
+        for (int i = size()-1; i>=0; i--) {
```

Comment:
1. Use spaces between operators.
2. Why reverse the iteration order?

Implementation discussion:
* We can use `boolean removed = removeIf(c::contains);` instead of the loop. However, this method uses the iterator and not random access, which *could* be less performant. The class implements `RandomAccess`, suggesting that this is the case, however, I don't see why. It seems that the implementation can be any list, such as `LinkedList`, which is not `RandomAccess`.
* Why do we need to override the inherited behavior? The super implementation, which is the one from `ModifiableObservableListBase`, does
  ```
        beginChange();
        try {
            boolean res = super.removeAll(c);
            return res;
        } finally {
            endChange();
        }
  ```
  (except that it does not check for empty collections at the start - should it?) The `super` here is `AbstractCollection`, which does `Iterator`-based removal.

Same goes for the other method.

Reply:
Rewriting backingList.removeIf
It is not used here because it requires doRemove processing.
This is the same as the original code.

The iteration order is reversed because index movement does not occur during deletion. This is the same as the original code.

Reviewers don't seem to be familiar with this code.

Reply:
wondering why the implementation with the BitSet was choosen at all? Certainly feels like a detour, so what are the corner cases that I'm too blind to see?

Reply:
I can't answer(, but my intuition is a pattern of lack of skill).
Officially, I don't know because the history is interrupted.

Reply:
digging a bit: the BitSet was introduced with https://bugs.openjdk.java.net/browse/JDK-8093144 to solve some problem with selectionModels - don't know whether those still hold (there had been extensive changes to selection since then).

Reply:
Thank you for searching for the origin of this strange code. However, as far as I read the comments, my intuition seems to be correct.

Reply:
I looked at the fix for [JDK-8093144](https://bugs.openjdk.java.net/browse/JDK-8093144), and the reason BitSet was introduced was to ensure that the elements are removed from this List in reverse order (prior to that fix, they were removed in forward order with the loop index being messed up). This patch preserves the correct behavior, and just looks to be a better fix for that earlier problem.

I do recommend running the failing test case from [JDK-8093144](https://bugs.openjdk.java.net/browse/JDK-8093144) to verify no regressions, but it looks like a good and safe fix to me.

@yososs I marked this discussion thread as unresolved mainly to make this comment, but also because you didn't fix the spacing suggested by @nlisker -- please do that.

Reply:
> the reason BitSet was introduced was to ensure that the elements are removed from this List in reverse order (prior to that fix, they were removed in forward order with the loop index being messed up).

But why do they need to be removed in reverse order to begin with? The super implementation does forward removal, or just use `removeIf`.

Reply:
It might not matter any more (presuming that it was done correctly), but it seems safer to leave the logic as-is to match the existing behavior.

Reply:
the problem was (and still is) in MultipleSelectionModelBase:

        selectedItems = new SelectedItemsReadOnlyObservableList<T>(selectedIndices, () -> getItemCount()) {
            @Override protected T getModelItem(int index) {
                return MultipleSelectionModelBase.this.getModelItem(index);
            }
        };

meaning that the selectedItems change during the removal of items (that's plain wrong!) which shows when removing the items from the start: 
       
        // removeAll before fix of 
        for (int i = 0; i < size(); ++i) {
            if (c.contains(get(i))) {
                remove(i);
                --i;
                modified = true;
            }
        }

doing so is basically valid (note the decremented i). Just the selectedItems look into the underlying model, so **c** is changing during the removal which is a no-go.

Returning to the old (pre [JDK-8093144](https://bugs.openjdk.java.net/browse/JDK-8093144)) still makes the tests against it fail (ListViewTest.test_rt35857 f.i.). 

Still wondering why the detour over the BitSet was choosen as fix (vs. the more natural remove-from-last). The listView test is passing for the bitSet and for the back-to-front approach. Can we imagine a context where the broken selectedItems impl would add wreckage to the latter?

Reply:
> The listView test is passing for the bitSet and for the back-to-front approach. Can we imagine a context where the broken selectedItems impl would add wreckage to the latter?

To answer my own question: yes. A failing test with the back-to-front approach (the existing test in ListViewTest modified in having the last item selected)

    @Test public void test_rt35857Back() {
        ObservableList<String> fxList = FXCollections.observableArrayList("A", "B", "C");
        final ListView<String> listView = new ListView<String>(fxList);
        
        listView.getSelectionModel().select(2);
        
        ObservableList<String> selectedItems = 
                listView.getSelectionModel().getSelectedItems();
        assertEquals(1, selectedItems.size());
        assertEquals("C", selectedItems.get(0));
        
        listView.getItems().removeAll(selectedItems);
        assertEquals(2, fxList.size());
        assertEquals("A", fxList.get(0));
        assertEquals("B", fxList.get(1));
    }
    
Feels like coming nearer to the why of the BitSet approach: it guards against the scenario where changing the current list implicitly changes the list given as parameter - in that case, it's unsafe to query the parameter list while removing items (c.contains simply reports nonsense).

This may happen whenever the parameter list does some kind of live lookup into the current list (such as f.i. SelectedItemsReadOnlyObservableList) - which is not as evil as I thought yesterday (did it myself in custom selectionModels ;) The BitSet solves it by a two-pass approach: first collect all items to remove/retain without (that is keep the parameter list in a valid state, allowing to safely access it), then do the removal without further accessing the (then invalid) parameter list. 

The fix at that time was a deliberate decision by the designer of the collections, so the context when it happens was deemed a valid use-case. Looks like we should **not** revert it.

Reply:
Nice catch! So now it's clear that the current approach was adopted because the source list itself can change as a result of removing an element from the target list in some cases, such as the one you pointed out above. I filed [JDK-8254040](https://bugs.openjdk.java.net/browse/JDK-8254040) to add the test case you listed above so we avoid a possible future regression.

So a two-pass algorithm is still needed: the first one to collect the elements to be removed, the second to actually remove them. While it isn't necessary to use a BitSet to collect the indexes to be removed, that does seems a reasonable approach. Unless there is a good reason to change it to some other two-pass algorithm, it's probably best to leave it as-is, in which case this PR and the associated JBS issue can be withdrawn.

Reply:
@kleopatra Thanks for the investigation. I had a feeling there is a practical reason and not a lack of skill :)

If there is no way to do removal simultaneously then we indeed need to keep the 2-pass approach.